swap/peer.go

```diff
@@ -0,0 +1,209 @@
+			// and we actually need to log this error as we are in an async routine; nobody is handling this error for now
+			log.Error("error submitting cheque", "err", err)
+			return
+		}
+		log.Debug("submit tx mined", "receipt", receipt)
+
+		receipt, err = otherSwap.CashChequeBeneficiary(opts, sp.backend, sp.swap.owner.Contract, big.NewInt(int64(actualAmount)))
+		if err != nil {
+			// TODO: do something with the error
+			// and we actually need to log this error as we are in an async routine; nobody is handling this error for now
+			log.Error("error cashing cheque", "err", err)
+			return
+		}
+		log.Debug("cash tx mined", "receipt", receipt)
+		log.Info("Cheque successfully submitted and cashed")
+	}()
+	return err
+}
+
+// processAndVerifyCheque verifies the cheque and compares it with the last received cheque
+// if the cheque is valid it will also be saved as the new last cheque
+func (sp *Peer) processAndVerifyCheque(cheque *Cheque) (uint64, error) {
+	if err := sp.verifyChequeProperties(cheque); err != nil {
+		return 0, err
+	}
+
+	lastCheque := sp.loadLastReceivedCheque()
+
+	// TODO: there should probably be a lock here?
+	expectedAmount, err := sp.swap.oracle.GetPrice(cheque.Honey)
+	if err != nil {
+		return 0, err
+	}
+
+	actualAmount, err := verifyChequeAgainstLast(cheque, lastCheque, expectedAmount)
+	if err != nil {
+		return 0, err
+	}
+
+	if err := sp.saveLastReceivedCheque(cheque); err != nil {
+		log.Error("error while saving last received cheque", "peer", sp.ID().String(), "err", err.Error())
+		// TODO: what do we do here? Related issue: https://github.com/ethersphere/swarm/issues/1515
+	}
+
+	return actualAmount, nil
+}
+
+// verifyChequeProperties verifies the signature and if the cheque fields are appropriate for this peer
+// it does not verify anything that requires knowing the previous cheque
+func (sp *Peer) verifyChequeProperties(cheque *Cheque) error {
+	if cheque.Contract != sp.contractAddress {
+		return fmt.Errorf("wrong cheque parameters: expected contract: %x, was: %x", sp.contractAddress, cheque.Contract)
+	}
+
+	// the beneficiary is the owner of the counterparty swap contract
+	if err := cheque.VerifySig(sp.beneficiary); err != nil {
+		return err
+	}
+
+	if cheque.Beneficiary != sp.swap.owner.address {
+		return fmt.Errorf("wrong cheque parameters: expected beneficiary: %x, was: %x", sp.swap.owner.address, cheque.Beneficiary)
+	}
+
+	if cheque.Timeout != 0 {
+		return fmt.Errorf("wrong cheque parameters: expected timeout to be 0, was: %d", cheque.Timeout)
+	}
+
+	return nil
+}
+
+// verifyChequeAgainstLast verifies that serial and amount are higher than in the previous cheque
+// furthermore it cheques that the increase in amount is as expected
+// returns the actual amount received in this cheque
+func verifyChequeAgainstLast(cheque *Cheque, lastCheque *Cheque, expectedAmount uint64) (uint64, error) {
+	actualAmount := cheque.Amount
+
+	if lastCheque != nil {
+		if cheque.Serial <= lastCheque.Serial {
```

Comment:
there should probably be some sort of protocol message dedicated to resolve such discrepancies (race conditions due to network conditions and so forth), but certainly not an issue for this PR